### PR TITLE
Add language settings / Apply language

### PR DESCRIPTION
**Pull request by Pluu**
## Issue
- #233

## Overview (Required)
- I restart to reflect the changed language. but, i think there is room for change that code.
- sorry, it's my first PR. Please look at the gif.

## Links
- Reference code : https://github.com/konifar/droidkaigi2016/blob/master/app/src/main/java/io/github/droidkaigi/confsched/fragment/SettingsFragment.java#L82

## Screenshot
Before | After
:--: | :--:
 |

### CODE REVIEW

File: app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java [hunk 1]
```diff
@@ -20,10 +25,51 @@
 
     public static final String LANG_EN = "en";
 
+    public static final List<String> SUPPORT_LANG = Arrays.asList(LANG_JA, LANG_EN);
```
Comment (konifar) [inline app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java]: Currently, we are going to manage languages by `Locale`, not `String`.
But it's okay to change it on another PR.

File: app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java [hunk 2]
```diff
@@ -20,10 +25,51 @@
...
+        Locale locale = new Locale(languageId);
+        Locale.setDefault(locale);
+        config.locale = locale;
+        context.getResources().updateConfiguration(config, context.getResources().getDisplayMetrics());
+    }
+
+    public static String getCurrentLanguageId(Context context) {
+        // This value would be stored language id or empty.
+        String languageId = DefaultPrefs.get(context).getLanguageId();
+        if (languageId.isEmpty()) {
```
Comment (konifar) [inline app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java]: `TextUtil.isEmpty(languageId)` is better if `languageId` may be null.

Comment (Pluu) [inline app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java]: Oh, Thank you for pointing out. :+1:

File: app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java (hunk 1 above)
Comment (Pluu) [inline app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java]:
```
public static final List<Locale> SUPPORT_LANG = Arrays.asList(Locale.JAPANESE, Locale.ENGLISH);
```
please check. This code is Ok?

**Comment by Pluu** [comment]
Fixed PR
- [x] LocaleUtils.java Conflicts
- [x] Language Manage Type (String -> Locale)

File: app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java [hunk 3]
```diff
@@ -1,25 +1,85 @@
...
 
 import io.github.droidkaigi.confsched2017.BuildConfig;
+import io.github.droidkaigi.confsched2017.R;
 import io.github.droidkaigi.confsched2017.pref.DefaultPrefs;
 import timber.log.Timber;
 
 public class LocaleUtil {
 
+    private static Locale DEFAULT_LANG = Locale.ENGLISH;
+    public static final List<Locale> SUPPORT_LANG = Arrays.asList(Locale.JAPANESE, Locale.ENGLISH);
```
Comment (konifar) [inline app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java]: 👍

File: app/src/main/java/io/github/droidkaigi/confsched2017/view/fragment/SettingsFragment.java [hunk 4]
```diff
@@ -49,4 +58,47 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle sa
...
+
+    @Override
+    public void showLanguagesDialog() {
+        List<Locale> locales = LocaleUtil.SUPPORT_LANG;
+        List<String> languages = Observable.fromIterable(locales)
+                .map(locale -> getString(LocaleUtil.getLanguage(locale)))
+                .toList()
+                .blockingGet();
+
+        List<String> languageIds = Observable.fromIterable(locales)
```
Comment (konifar) [inline app/src/main/java/io/github/droidkaigi/confsched2017/view/fragment/SettingsFragment.java]: `Observable` is a little slow. How about using `Stream`?

File: app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java [hunk 5]
```diff
@@ -1,25 +1,85 @@
...
+
+    public static String getCurrentLanguage(Context context) {
+        return context.getString(getLanguage(LocaleUtil.getCurrentLanguageId(context)));
+    }
+
+    public static @StringRes int getLanguage(@NonNull Locale locale) {
+        return getLanguage(getLocaleLanguageId(locale));
+    }
+
+    public static @StringRes int getLanguage(@NonNull String languageId) {
```
Comment (konifar) [inline app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java]: I prefer to put annotations above the method like this.

```java
@StringRes
public static int getLanguage(@NonNull String languageId) {
```

File: app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java [hunk 6]
```diff
@@ -1,25 +1,85 @@
...
 import java.util.TimeZone;
 
 import io.github.droidkaigi.confsched2017.BuildConfig;
+import io.github.droidkaigi.confsched2017.R;
 import io.github.droidkaigi.confsched2017.pref.DefaultPrefs;
 import timber.log.Timber;
 
 public class LocaleUtil {
 
+    private static Locale DEFAULT_LANG = Locale.ENGLISH;
```
Comment (jmatsu) [inline app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java]: should be `final` if possible. 😃

Comment (Pluu) [inline app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java]: :+1: Thank you!! Next PR Apply.

File: app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java [hunk 7]
```diff
@@ -1,25 +1,85 @@
...
+    public static void initLocale(Context context) {
+        setLocale(context, getCurrentLanguageId(context));
+    }
+
+    public static void setLocale(Context context, String languageId) {
+        Configuration config = new Configuration();
+        DefaultPrefs.get(context).putLanguageId(languageId);
+        Locale locale = new Locale(languageId);
+        Locale.setDefault(locale);
+        config.locale = locale;
```
Comment (jmatsu) [inline app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java]: This has been deprecated in API 24.
https://developer.android.com/reference/android/content/res/Configuration.html#locale

updateConfiguration has been also deprecated in API 25.
https://developer.android.com/reference/android/content/res/Resources.html#updateConfiguration(android.content.res.Configuration, android.util.DisplayMetrics)

Could you try new way to apply Locale?

File: app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java (hunk 5 above)
Comment (Pluu) [inline app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java]: OK. Next PR. This Style Apply!

File: app/src/main/java/io/github/droidkaigi/confsched2017/view/fragment/SettingsFragment.java (hunk 4 above)
Comment (Pluu) [inline app/src/main/java/io/github/droidkaigi/confsched2017/view/fragment/SettingsFragment.java]: I did not know the existence of `com.annimon:stream:1.1.5`.
NEXT PR. `Observable` to `Stream`.

File: app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java (hunk 7 above)
Comment (konifar) [inline app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java]: Good point! 👍 Thanks @jmatsu

**Comment by Pluu** [comment]
@kgmyshin thank you report. Please tell me the Device information.

- Default language
- Android Software Version
- etc, more infomation

**Comment by konifar** [comment]
@Pluu Maybe it doesn't depend on device spec.

`LocaleUtil.getLanguage(locale)` returns language name from strings.xml.
https://github.com/DroidKaigi/conference-app-2017/pull/270/files#diff-4edd5c3319276c7a31c45f946ee9ff9fR68

On English locale setting, this `languages` is set `[English, Japanese]`.
But it should be set `[English, 日本語]`.
So I think you have to fix `Locale.getLanguage(locale)` logic or create another method.

**Comment by Pluu** [comment]
my think was wrong.
thought that the display would change depending on the language.

`I will fix it now`

**Comment by konifar** [comment]
Sorry, the conflict occurred... Can you merge master?

File: app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java (hunk 7 above)
Comment (Pluu) [inline app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java]: 👍 Good

**Comment by Pluu** [comment]
@konifar After `Configuration` corrections, and PR again. now, testing progress in `API 25`.

**Comment by Pluu** [comment]
ja | en
:--: | :--:
 | 

## Fixed

- Display language dialog
- if/else Configuration.html#locale

## Not Fixed

- updateConfiguration has been also deprecated in API 25

**Comment by konifar** [comment]
Good! 👍 

> updateConfiguration has been also deprecated in API 25

Do you want to fix it on this PR? or another PR?

File: app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java (hunk 7 above)
Comment (Pluu) [inline app/src/main/java/io/github/droidkaigi/confsched2017/util/LocaleUtil.java]: current PR. 
- Fixed, if/else Configuration.html#locale
- Not Fixed, `updateConfiguration has been also deprecated in API 25`

**Comment by Pluu** [comment]
@konifar fix on this PR.

**Comment by konifar** [comment]
LGTM! Thanks 👍

**Pull request merged**